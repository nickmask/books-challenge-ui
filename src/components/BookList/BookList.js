import React from "react";

import BookCard from "../BookCard/BookCard";
import LoadingSpinner from "../LoadingSpinner/LoadingSpinner";
import { fetchBooks } from "../../utils/api";

import styles from "./BookList.module.css";

class BookList extends React.Component {
  state = {
    books: null,
    isFetching: true,
    error: null
  };

  async componentDidMount() {
    if (!this.state.books) {
      try {
        const books = await fetchBooks();
        this.setState({ isFetching: false, books });
      } catch (error) {
        alert("Error getting books. Please try again later");
        this.setState({
          error
        });
      }
    }
  }

  render() {
    const { isFetching, books } = this.state;
    return (
      <div className={styles.container}>
        {isFetching ? (
          <LoadingSpinner />
        ) : (
          <>
            <h1 className={styles.pageHeader}>Books</h1>
            <hr className={styles.separator} />
            <div className={styles.list}>
              {books.map(book => (
                <BookCard
                  key={book.id}
                  title={book.title}
                  author={`${book.author.firstName} ${book.author.lastName}`}
                  price={book.price}
                  description={book.description}
                  stockAmount={book.stockAmount}
                  thumbnail={book.thumbnail}
                  id={book.id}
                />
              ))}
            </div>
          </>
        )}
      </div>
    );
  }
}

export default BookList;
