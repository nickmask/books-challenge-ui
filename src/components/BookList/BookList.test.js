import React from "react";
import { shallow } from "enzyme";
import createRouterContext from "react-router-test-context";

import BookList from "./BookList";

describe("BookList", () => {
  const context = createRouterContext();

  const books = {
    results: [
      {
        id: "c1dc42ab-548f-4f96-8350-2f56303d7532",
        title: "A Life Changed",
        author: {
          id: "b6545db7-2a50-4532-8225-3fec60470502",
          firstName: "John",
          lastName: "Doe"
        },
        price: 10,
        description:
          "Nulla sunt sunt pariatur officia et esse fugiat proident est. Ad laboris ullamco adipisicing nisi in ex. Magna voluptate culpa velit exercitation nisi ipsum est do.",
        stockAmount: 0,
        thumbnail:
          "https://images.unsplash.com/photo-1496484805162-42c00f79cdce?dpr=1&auto=format&fit=crop&w=568&h=379&q=60&cs=tinysrgb&ixid=dW5zcGxhc2guY29tOzs7Ozs%3D",
        isbn: "0000044280374"
      },
      {
        id: "da435429-c698-4b5b-9865-808995d82e47",
        title: "The Math Behind React",
        author: {
          id: "1335ccbe-04b6-4d1a-a339-b9ee6202847a",
          firstName: "Jane",
          lastName: "Langton"
        },
        price: 48,
        description:
          "Nulla sunt sunt pariatur officia et esse fugiat proident est. Ad laboris ullamco adipisicing nisi in ex. Magna voluptate culpa velit exercitation nisi ipsum est do.",
        stockAmount: 2,
        thumbnail:
          "https://images.unsplash.com/photo-1453733190371-0a9bedd82893?dpr=1&auto=format&fit=crop&w=568&h=426&q=60&cs=tinysrgb&ixid=dW5zcGxhc2guY29tOzs7Ozs%3D",
        isbn: "0000036338645"
      }
    ]
  };

  beforeEach(() => {
    const mockJsonPromise = Promise.resolve(books);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise
    });
    jest.spyOn(global, "fetch").mockImplementation(() => mockFetchPromise);
  });

  afterEach(() => {
    fetch.mockReset();
  });

  it("should render correctly", async () => {
    const component = await shallow(<BookList />, { context });

    await new Promise(resolve => setTimeout(resolve, 0)); // Re-render is not reflected in component without this

    expect(component).toMatchSnapshot();
  });

  it("should render correct number of BookCards", async () => {
    const component = await shallow(<BookList />, { context });

    await new Promise(resolve => setTimeout(resolve, 0)); // Re-render is not reflected in component without this

    const cards = component.find("BookCard");

    expect(cards.length).toEqual(2);
  });

  it("should trigger a fetch call to get books", async () => {
    const component = await shallow(<BookList />, { context });

    await new Promise(resolve => setTimeout(resolve, 0)); // Re-render is not reflected in component without this

    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch).toHaveBeenCalledWith(
      "https://booksontap.azurewebsites.net/api/books"
    );
  });
});
