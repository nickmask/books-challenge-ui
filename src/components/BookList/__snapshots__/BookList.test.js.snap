// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookList should render correctly 1`] = `
<div
  className="container"
>
  <h1
    className="pageHeader"
  >
    Books
  </h1>
  <hr
    className="separator"
  />
  <div
    className="list"
  >
    <BookCard
      author="John Doe"
      description="Nulla sunt sunt pariatur officia et esse fugiat proident est. Ad laboris ullamco adipisicing nisi in ex. Magna voluptate culpa velit exercitation nisi ipsum est do."
      id="c1dc42ab-548f-4f96-8350-2f56303d7532"
      key="c1dc42ab-548f-4f96-8350-2f56303d7532"
      price={10}
      stockAmount={0}
      thumbnail="https://images.unsplash.com/photo-1496484805162-42c00f79cdce?dpr=1&auto=format&fit=crop&w=568&h=379&q=60&cs=tinysrgb&ixid=dW5zcGxhc2guY29tOzs7Ozs%3D"
      title="A Life Changed"
    />
    <BookCard
      author="Jane Langton"
      description="Nulla sunt sunt pariatur officia et esse fugiat proident est. Ad laboris ullamco adipisicing nisi in ex. Magna voluptate culpa velit exercitation nisi ipsum est do."
      id="da435429-c698-4b5b-9865-808995d82e47"
      key="da435429-c698-4b5b-9865-808995d82e47"
      price={48}
      stockAmount={2}
      thumbnail="https://images.unsplash.com/photo-1453733190371-0a9bedd82893?dpr=1&auto=format&fit=crop&w=568&h=426&q=60&cs=tinysrgb&ixid=dW5zcGxhc2guY29tOzs7Ozs%3D"
      title="The Math Behind React"
    />
  </div>
</div>
`;
